core/modules/wbase/SendChannelShared.h

```diff
@@ -0,0 +1,161 @@
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <http://www.lsstcorp.org/LegalNotices/>.
+ */
+
+#ifndef LSST_QSERV_WBASE_SENDCHANNELSHARED_H
+#define LSST_QSERV_WBASE_SENDCHANNELSHARED_H
+
+// System headers
+#include <functional>
+#include <memory>
+#include <queue>
+#include <string>
+#include <stdexcept>
+#include <thread>
+
+// Qserv headers
+#include "global/Bug.h"
+#include "wbase/SendChannel.h"
+#include "wbase/TransmitData.h"
+#include "wcontrol/TransmitMgr.h"
+
+namespace lsst {
+namespace qserv {
+namespace wbase {
+
+/// A class that provides a SendChannel object with synchronization so it can be
+/// shared by across multiple threads. Due to what may be sent, the synchronization locking
+/// is needs to be available outside of the class.
+class SendChannelShared {
+public:
+    using Ptr = std::shared_ptr<SendChannelShared>;
+
+    /// To help ensure that _streamMutex is locked before calling,
+    /// many member functions require a StreamGuard argument.
+    using StreamGuard = std::lock_guard<std::mutex> const&;
+
+    SendChannelShared()=delete;
+    SendChannelShared(SendChannelShared const&) = delete;
+    SendChannelShared& operator=(SendChannelShared const&) = delete;
+
+    static Ptr create(SendChannel::Ptr const& sendChannel, wcontrol::TransmitMgr::Ptr const& transmitMgr);
+
+    ~SendChannelShared() = default;
+
+    /// Wrappers for SendChannel public functions that may need to be used
+    /// by threads.
+    /// @see SendChannel::send
+    bool send(StreamGuard sLock, char const* buf, int bufLen) {
+        return _sendChannel->send(buf, bufLen);
+    }
+
+    /// @see SendChannel::sendError
+    bool sendError(StreamGuard sLock, std::string const& msg, int code) {
+        return _sendChannel->sendError(msg, code);
+    }
+
+    /// @see SendChannel::sendFile
+    bool sendFile(StreamGuard sLock, int fd, SendChannel::Size fSize) {
+        return _sendChannel->sendFile(fd, fSize);
+    }
+
+    /// @see SendChannel::sendStream
+    bool sendStream(StreamGuard sLock, xrdsvc::StreamBuffer::Ptr const& sBuf, bool last) {
+        return _sendChannel->sendStream(sBuf, last);
+    }
+
+    /// @see SendChannel::kill
+    bool kill(StreamGuard sLock);
+
+    /// @see SendChannel::isDead
+    bool isDead() { return _sendChannel->isDead(); }
+
+
+    /// Set the number of Tasks that will be sent using this SendChannel.
+    /// This should not be changed once set.
+    void setTaskCount(int taskCount);
+
+
+    /// Try to transmit the data in tData.
+    /// If the queue already has at least 2 TransmitData objects, addTransmit
+    /// may wait before returning. It's more efficient use of memory to
+    /// collect results from MariaDB as they are sent to the czar than
+    /// to read them all in at once.
+    bool addTransmit(bool cancelled, bool erred, bool last, bool largeResult,
```

Comment:
OK, the comment doesn't really tell me how this interplays with removing the thread. Not sure what the comment really has to say but given your changes it makes me anyone else looking at the code wonder.